# Ch.11.(Part 2) Deep learning for text

Chapter contents:

* Transformer 아키텍처
* Sequence-to-sequence 학습


## Ch.11.3 The Transformer architecture

2017년부터 새로운 모델 아키텍처: Transformer(Attention is all you need)

"neural attention"이라는 간단한 메커니즘을 사용하여 순환(recurrent) 레이어나 컨볼루션 레이어가 없는 강력한 시퀀스 모델을 만들 수 있습니다.

### Understanding self-attention

모델이 보는 모든 입력 정보가 당면한 작업에 똑같이 중요하지는 않습니다.

모델은 일부 특징에 "더 많은 주의를 기울이고(pay more attention)" 다른 특징에 "덜 주의를 기울여야(pay less attention)" 합니다.

비슷한 생각:

*  convnet의 Max pooling은 공간 영역의 특징 풀을 살펴보고 유지할 특징을 하나만 선택합니다.
* TF-IDF 정규화는 다른 토큰이 전달할 가능성이 있는 정보의 양에 따라 토큰에 중요도 점수를 할당합니다.

<img src="https://drek4537l1klr.cloudfront.net/chollet2/Figures/11-05.png" width="300"><p style="text-align:center">Figure 11.5 The general concept of “attention” in deep learning: input features get assigned “attention scores,” which can be used to inform the next representation of the input. From raw text to vectors.</p>

Attention 메커니즘을 사용하여 특징을 컨텍스트-인식(context-aware)이 가능하게 할 수 있습니다.

벡터 공간은 서로 다른 단어 간의 의미 관계의 "모양"을 캡처합니다.

한 단어는 고정된 위치를 갖는다

그러나: 단어의 의미는 일반적으로 문맥에 따라 다릅니다.

스마트 임베딩 공간: 단어를 둘러싼 다른 단어에 따라 단어에 대해 다른 벡터 표현 제공

self-attention의 목적은 시퀀스에서 관련 토큰의 표현을 사용하여 토큰의 표현을 변조하는 것입니다.

<img src="https://drek4537l1klr.cloudfront.net/chollet2/Figures/11-06.png" width="500"><p style="text-align:center">Figure 11.6. Self-attention: attention scores are computed between “station” and every other word in the sequence, and they are then used to weight a sum of word vectors that becomes the new “station” vector.</p>

* Compute relevancy scores (dot product)
* Scale (1/sqrt) and softmax
* Weighted (scores) sum of vectors

In [ ]:
def self_attention(input_sequence):
    output = np.zeros(shape=input_sequence.shape)
    for i, pivot_vector in enumerate(input_sequence):
        scores = np.zeros(shape=(len(input_sequence),))
        for j, vector in enumerate(input_sequence):
            scores[j] = np.dot(pivot_vector, vector.T)
        scores /= np.sqrt(input_sequence.shape[1])
        scores = softmax(scores)
        new_pivot_representation = np.zeros(shape=pivot_vector.shape)
        for j, vector in enumerate(input_sequence):
            new_pivot_representation += vector * scores[j]
        output[i] = new_pivot_representation
    return output

#### Generalized self-attention: the query-key-value model

self-attention을 위한 백터화 구현
```
    num_heads = 4 
    embed_dim = 256 
    mha_layer = MultiHeadAttention(num_heads=num_heads, key_dim=embed_dim)
    outputs = mha_layer(inputs, inputs, inputs)
```
지금까지 단 하나의 시퀀스를 사용함. 
그러나 기계 번역용으로 개발된 Transformer (2개의 시퀀스: source, target)

<img src="https://drek4537l1klr.cloudfront.net/chollet2/Figures/11-06-UN01.png" width="450">

* 입력(A)의 각 토큰에 대해, 
* 토큰이 입력(B)의 모든 토큰과 얼마나 관련되어 있는지 계산
* 이 점수를 사용하여 입력(C)의 토큰 합계에 가중치를 부여합니다.

<img src="https://drek4537l1klr.cloudfront.net/chollet2/Figures/11-06-UN02.png" width="450">
 
<img src="https://drek4537l1klr.cloudfront.net/chollet2/Figures/11-07.png" width="500"><p style="text-align:center">Figure 11.7 Retrieving images from a database: the “query” is compared to a set of “keys,” and the match scores are used to rank “values” (images).</p>

* Source: "How's the weather today?"
* Target: "¿Qué tiempo hace hoy?" 

* query: target
* keys: source
* values: source

### Multi-head attention

다중 헤드(Multi-head): self-attention layer의 출력 공간은 개별적으로 학습된 독립적인 부분 공간 세트로 나누어집니다.

초기 쿼리(query), 키(key) 및 값(value)은 3개의 독립적인 dense projections 세트를 통해 전송되어 3개의 개별 벡터가 생성됩니다.

각 벡터는 neural attention 을 통해 처리됩니다.

3개의 출력은 다시 하나의 출력 시퀀스로 연결됩니다.

학습 가능한 dense projections의 존재는 레이어가 무언가를 학습할 수 있도록 합니다(순수한 상태없는 변환(stateless transformation)과 반대).

독립 헤드(Independent heads): 레이어가 각 토큰에 대한 다양한 특징 그룹을 학습하도록 돕습니다.

<img src="https://drek4537l1klr.cloudfront.net/chollet2/Figures/11-08.png" width="400"><p style="text-align:center">Figure 11.8 The MultiHeadAttention layer</p>

### The Transformer encoder

Dense projections 작업 -> attention 메커니즘의 출력에도 추가함.

또한

* residual connection 추가(깊어지기 때문에)
* 정규화 추가(역전파 동안 더 나은 gradients 흐름을 위해)

<img src="https://drek4537l1klr.cloudfront.net/chollet2/Figures/11-09.png" width="200"><p style="text-align:center">Figure 11.9 The TransformerEncoder chains a MultiHeadAttention layer with a dense projection and adds normalization as well as residual connections.</p>

**Getting the data**

In [ ]:
# !curl -O https://ai.stanford.edu/~amaas/data/sentiment/aclImdb_v1.tar.gz
# !tar -xf aclImdb_v1.tar.gz
# !rm -r aclImdb/train/unsup

**Preparing the data**

In [0]:
# import os, pathlib, shutil, random
# from tensorflow import keras
# batch_size = 32
# base_dir = pathlib.Path("aclImdb")
# val_dir = base_dir / "val"
# train_dir = base_dir / "train"
# for category in ("neg", "pos"):
#     os.makedirs(val_dir / category)
#     files = os.listdir(train_dir / category)
#     random.Random(1337).shuffle(files)
#     num_val_samples = int(0.2 * len(files))
#     val_files = files[-num_val_samples:]
#     for fname in val_files:
#         shutil.move(train_dir / category / fname,
#                     val_dir / category / fname)

In [4]:
from tensorflow import keras
batch_size = 32
train_ds = keras.utils.text_dataset_from_directory(
    "aclImdb/train", batch_size=batch_size
)
val_ds = keras.utils.text_dataset_from_directory(
    "aclImdb/val", batch_size=batch_size
)
test_ds = keras.utils.text_dataset_from_directory(
    "aclImdb/test", batch_size=batch_size
)
text_only_train_ds = train_ds.map(lambda x, y: x)

Found 20000 files belonging to 2 classes.
Found 5000 files belonging to 2 classes.
Found 25000 files belonging to 2 classes.


**Vectorizing the data**

In [5]:
from tensorflow.keras import layers

max_length = 600
max_tokens = 20000
text_vectorization = layers.TextVectorization(
    max_tokens=max_tokens,
    output_mode="int",
    output_sequence_length=max_length,
)
text_vectorization.adapt(text_only_train_ds)

int_train_ds = train_ds.map(
    lambda x, y: (text_vectorization(x), y),
    num_parallel_calls=4)
int_val_ds = val_ds.map(
    lambda x, y: (text_vectorization(x), y),
    num_parallel_calls=4)
int_test_ds = test_ds.map(
    lambda x, y: (text_vectorization(x), y),
    num_parallel_calls=4)

In [12]:
list(int_test_ds.take(1))[0]

(<tf.Tensor: shape=(32, 600), dtype=int64, numpy=
 array([[    4,   473,     5, ...,     0,     0,     0],
        [ 9793,     7,   445, ...,     0,     0,     0],
        [  615,    10,  2449, ...,     0,     0,     0],
        ...,
        [14541,     9,     7, ...,     0,     0,     0],
        [   10,   118,    12, ...,     0,     0,     0],
        [ 9493,     2,  1290, ...,     0,     0,     0]], dtype=int64)>,
 <tf.Tensor: shape=(32,), dtype=int32, numpy=
 array([0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 1, 1, 1, 1,
        1, 1, 0, 1, 1, 1, 0, 1, 0, 1])>)

**Transformer encoder implemented as a subclassed `Layer`**

In [6]:
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers

class TransformerEncoder(layers.Layer):
    def __init__(self, embed_dim, dense_dim, num_heads, **kwargs):
        super().__init__(**kwargs)
        self.embed_dim = embed_dim
        self.dense_dim = dense_dim
        self.num_heads = num_heads
        self.attention = layers.MultiHeadAttention(
            num_heads=num_heads, key_dim=embed_dim)
        self.dense_proj = keras.Sequential(
            [layers.Dense(dense_dim, activation="relu"),
             layers.Dense(embed_dim),]
        )
        self.layernorm_1 = layers.LayerNormalization()
        self.layernorm_2 = layers.LayerNormalization()

    def call(self, inputs, mask=None):
        if mask is not None:
            mask = mask[:, tf.newaxis, :]
        attention_output = self.attention(
            inputs, inputs, attention_mask=mask)
        proj_input = self.layernorm_1(inputs + attention_output)
        proj_output = self.dense_proj(proj_input)
        return self.layernorm_2(proj_input + proj_output)

    def get_config(self):
        config = super().get_config()
        config.update({
            "embed_dim": self.embed_dim,
            "num_heads": self.num_heads,
            "dense_dim": self.dense_dim,
        })
        return config

**Using the Transformer encoder for text classification**

In [7]:
vocab_size = 20000
embed_dim = 256
num_heads = 2
dense_dim = 32

inputs = keras.Input(shape=(None,), dtype="int64")
x = layers.Embedding(vocab_size, embed_dim)(inputs)
x = TransformerEncoder(embed_dim, dense_dim, num_heads)(x)
x = layers.GlobalMaxPooling1D()(x)
x = layers.Dropout(0.5)(x)
outputs = layers.Dense(1, activation="sigmoid")(x)
model = keras.Model(inputs, outputs)
model.compile(optimizer="rmsprop",
              loss="binary_crossentropy",
              metrics=["accuracy"])
model.summary()

Model: "model"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 input_1 (InputLayer)        [(None, None)]            0         
                                                                 
 embedding (Embedding)       (None, None, 256)         5120000   
                                                                 
 transformer_encoder (Transf  (None, None, 256)        543776    
 ormerEncoder)                                                   
                                                                 
 global_max_pooling1d (Globa  (None, 256)              0         
 lMaxPooling1D)                                                  
                                                                 
 dropout (Dropout)           (None, 256)               0         
                                                                 
 dense_2 (Dense)             (None, 1)                 257   

**Training and evaluating the Transformer encoder based model**

In [8]:
callbacks = [
    keras.callbacks.ModelCheckpoint("transformer_encoder.keras",
                                    save_best_only=True)
]
model.fit(int_train_ds, validation_data=int_val_ds, epochs=20, callbacks=callbacks)
model = keras.models.load_model(
    "transformer_encoder.keras",
    custom_objects={"TransformerEncoder": TransformerEncoder})
print(f"Test acc: {model.evaluate(int_test_ds)[1]:.3f}")

Epoch 1/20
625/625 [==============================] - 37s 51ms/step - loss: 0.4791 - accuracy: 0.7758 - val_loss: 0.3446 - val_accuracy: 0.8504
Epoch 2/20
625/625 [==============================] - 30s 49ms/step - loss: 0.3156 - accuracy: 0.8662 - val_loss: 0.3205 - val_accuracy: 0.8640
Epoch 3/20
625/625 [==============================] - 30s 49ms/step - loss: 0.2486 - accuracy: 0.8988 - val_loss: 0.2945 - val_accuracy: 0.8852
Epoch 4/20
625/625 [==============================] - 31s 49ms/step - loss: 0.1933 - accuracy: 0.9245 - val_loss: 0.3295 - val_accuracy: 0.8874
Epoch 5/20
625/625 [==============================] - 31s 49ms/step - loss: 0.1534 - accuracy: 0.9421 - val_loss: 0.3528 - val_accuracy: 0.8836
Epoch 6/20
625/625 [==============================] - 31s 50ms/step - loss: 0.1268 - accuracy: 0.9533 - val_loss: 0.3746 - val_accuracy: 0.8910
Epoch 7/20
625/625 [==============================] - 32s 51ms/step - loss: 0.1124 - accuracy: 0.9577 - val_loss: 0.3498 - val_accuracy:

* binary_1gram: 0.887
* binary_2gram: 0.892
* tfidf_2gram: 0.897
* one_hot_bidir_lstm: 0.873
* embeddings_bidir_lstm: 0.868
* embeddings_bidir_lstm_with_masking: 0.872
* glove_embeddings_sequence model: 0.868
* transformer_encoder: 0.874

기본 모델보다 좋지 않습니다. 왜요?

현재 Transformer 모델: 시퀀스 모델이 아닙니다.

Attention 계층: 토큰을 집합(set)으로 봅니다.

토큰 순서 변경: 동일한 컨텍스트 인식 표현

<img src="https://drek4537l1klr.cloudfront.net/chollet2/Figures/11-10.png" width="250"><p style="text-align:center">Figure 11.10 Features of different types of NLP models</p>

#### Using positional encoding to re-inject order information


위치 인코딩: 모델에 단어-순서 정보에 대한 액세스 부여

각 단어 임베딩에 단어의 위치 추가

워드 임베딩 구성요소:
* 단어 벡터(word vector)(독립 표현)
* 위치 벡터(position vector)(단어의 위치를 나타냄)

구현

모델은 위치-임베딩(position-embedding) 벡터를 학습합니다(단어 인덱스를 포함하는 것과 동일한 방식).

**Implementing positional embedding as a subclassed layer**

In [9]:
class PositionalEmbedding(layers.Layer):
    def __init__(self, sequence_length, input_dim, output_dim, **kwargs):
        super().__init__(**kwargs)
        self.token_embeddings = layers.Embedding(
            input_dim=input_dim, output_dim=output_dim)
        self.position_embeddings = layers.Embedding(
            input_dim=sequence_length, output_dim=output_dim)
        self.sequence_length = sequence_length
        self.input_dim = input_dim
        self.output_dim = output_dim

    def call(self, inputs):
        length = tf.shape(inputs)[-1]
        positions = tf.range(start=0, limit=length, delta=1)
        embedded_tokens = self.token_embeddings(inputs)
        embedded_positions = self.position_embeddings(positions)
        return embedded_tokens + embedded_positions

    def compute_mask(self, inputs, mask=None):
        return tf.math.not_equal(inputs, 0)

    def get_config(self):
        config = super().get_config()
        config.update({
            "output_dim": self.output_dim,
            "sequence_length": self.sequence_length,
            "input_dim": self.input_dim,
        })
        return config

#### Putting it all together: A text-classification Transformer

**Combining the Transformer encoder with positional embedding**

In [10]:
vocab_size = 20000
sequence_length = 600
embed_dim = 256
num_heads = 2
dense_dim = 32

inputs = keras.Input(shape=(None,), dtype="int64")
x = PositionalEmbedding(sequence_length, vocab_size, embed_dim)(inputs)
x = TransformerEncoder(embed_dim, dense_dim, num_heads)(x)
x = layers.GlobalMaxPooling1D()(x)
x = layers.Dropout(0.5)(x)
outputs = layers.Dense(1, activation="sigmoid")(x)
model = keras.Model(inputs, outputs)
model.compile(optimizer="rmsprop",
              loss="binary_crossentropy",
              metrics=["accuracy"])
model.summary()

callbacks = [
    keras.callbacks.ModelCheckpoint("full_transformer_encoder.keras",
                                    save_best_only=True)
]
model.fit(int_train_ds, validation_data=int_val_ds, epochs=20, callbacks=callbacks)
model = keras.models.load_model(
    "full_transformer_encoder.keras",
    custom_objects={"TransformerEncoder": TransformerEncoder,
                    "PositionalEmbedding": PositionalEmbedding})
print(f"Test acc: {model.evaluate(int_test_ds)[1]:.3f}")

Model: "model_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 input_2 (InputLayer)        [(None, None)]            0         
                                                                 
 positional_embedding (Posit  (None, None, 256)        5273600   
 ionalEmbedding)                                                 
                                                                 
 transformer_encoder_1 (Tran  (None, None, 256)        543776    
 sformerEncoder)                                                 
                                                                 
 global_max_pooling1d_1 (Glo  (None, 256)              0         
 balMaxPooling1D)                                                
                                                                 
 dropout_1 (Dropout)         (None, 256)               0         
                                                           

* binary_1gram: 0.887
* binary_2gram: 0.892
* tfidf_2gram: 0.897
* one_hot_bidir_lstm: 0.873
* embeddings_bidir_lstm: 0.868
* embeddings_bidir_lstm_with_masking: 0.872
* glove_embeddings_sequence model: 0.868
* transformer_encoder: 0.874
* full_transformer_encoder: 0.882

Improved but still not as good as the Bag-of-words ones.

### When to use sequence models over bag-of-words models?

Bag-of-words는 구식이 아닙니다.

어떤 경우에는 bag-of-bigrams 위에 Dense layers의 작은 스택이 완벽하게 유효합니다.

다양한 유형의 텍스트 데이터 세트에 대한 다양한 텍스트 분류 기술 분석

중요 기준: 훈련 데이터의 샘플 수와 샘플당 평균 단어 수 간의 비율


<img src="https://drek4537l1klr.cloudfront.net/chollet2/Figures/11-11.png" width="350"><p style="text-align:center">Figure 11.11 A simple heuristic for selecting a text-classification model: the ratio between the number of training samples and the mean number of words per sample</p>

Sequence models work best when

* lots of training data is available
* each sample is relatively short

## Beyond text classification: Sequence-to-sequence learning

Sequence-to-sequence learning applications

* Machine translation
* Text summarization
* Chatbots
* Text generation


General template:
 * Encoder: turns the source sequence into an intermediate representation
 * Decoder: trained to predict the next token i in the target sequence by looking at
     * previous tokens (0 to i-1) and
     * the encoded source sequence
     
During Inference: no access to target sequence. Generate it one token at a time

1. obtain the encoded source sequence from the encoder
2. decoder starts with:
  * encoded source sequence
  * initial "seed" token [start]
  
  predicts the first real token in the sequence
  
3. Predicted sequences fed back into the decoder (until stop token [end] is generated)

<img src="https://drek4537l1klr.cloudfront.net/chollet2/Figures/11-12.png" width="450"><p style="text-align:center">Figure 11.12 Sequence-to-sequence learning: the source sequence is processed by the encoder and is then sent to the decoder. The decoder looks at the target sequence so far and predicts the target sequence offset by one step in the future. During inference, we generate one target token at a time and feed it back into the decoder.</p>

### A machine translation example

In [14]:
# !wget http://storage.googleapis.com/download.tensorflow.org/data/spa-eng.zip
!unzip -q spa-eng.zip

In [16]:
text_file = "spa-eng/spa.txt"
with open(text_file, encoding='utf-8') as f:
    lines = f.read().split("\n")[:-1]
text_pairs = []
for line in lines:
    english, spanish = line.split("\t")
    spanish = "[start] " + spanish + " [end]"
    text_pairs.append((english, spanish))

In [17]:
import random
print(random.choice(text_pairs))

('Tom waited patiently for Mary.', '[start] Tom esperó pacientemente a Mary. [end]')


In [18]:
import random
random.shuffle(text_pairs)
num_val_samples = int(0.15 * len(text_pairs))
num_train_samples = len(text_pairs) - 2 * num_val_samples
train_pairs = text_pairs[:num_train_samples]
val_pairs = text_pairs[num_train_samples:num_train_samples + num_val_samples]
test_pairs = text_pairs[num_train_samples + num_val_samples:]

**Vectorizing the English and Spanish text pairs**

In [19]:
import tensorflow as tf
import string
import re

strip_chars = string.punctuation + "¿"
strip_chars = strip_chars.replace("[", "")
strip_chars = strip_chars.replace("]", "")

def custom_standardization(input_string):
    lowercase = tf.strings.lower(input_string)
    return tf.strings.regex_replace(
        lowercase, f"[{re.escape(strip_chars)}]", "")

vocab_size = 15000
sequence_length = 20

source_vectorization = layers.TextVectorization(
    max_tokens=vocab_size,
    output_mode="int",
    output_sequence_length=sequence_length,
)
target_vectorization = layers.TextVectorization(
    max_tokens=vocab_size,
    output_mode="int",
    output_sequence_length=sequence_length + 1,
    standardize=custom_standardization,
)
train_english_texts = [pair[0] for pair in train_pairs]
train_spanish_texts = [pair[1] for pair in train_pairs]
source_vectorization.adapt(train_english_texts)
target_vectorization.adapt(train_spanish_texts)

**Preparing datasets for the translation task**

In [20]:
batch_size = 64

def format_dataset(eng, spa):
    eng = source_vectorization(eng)
    spa = target_vectorization(spa)
    return ({
        "english": eng,
        "spanish": spa[:, :-1],
    }, spa[:, 1:])

def make_dataset(pairs):
    eng_texts, spa_texts = zip(*pairs)
    eng_texts = list(eng_texts)
    spa_texts = list(spa_texts)
    dataset = tf.data.Dataset.from_tensor_slices((eng_texts, spa_texts))
    dataset = dataset.batch(batch_size)
    dataset = dataset.map(format_dataset, num_parallel_calls=4)
    return dataset.shuffle(2048).prefetch(16).cache()

train_ds = make_dataset(train_pairs)
val_ds = make_dataset(val_pairs)

In [21]:
for inputs, targets in train_ds.take(1):
    print(f"inputs['english'].shape: {inputs['english'].shape}")
    print(f"inputs['spanish'].shape: {inputs['spanish'].shape}")
    print(f"targets.shape: {targets.shape}")

inputs['english'].shape: (64, 20)
inputs['spanish'].shape: (64, 20)
targets.shape: (64, 20)


### Sequence-to-sequence learning with RNNs

**GRU-based encoder**

In [22]:
from tensorflow import keras
from tensorflow.keras import layers

embed_dim = 256
latent_dim = 1024

source = keras.Input(shape=(None,), dtype="int64", name="english")
x = layers.Embedding(vocab_size, embed_dim, mask_zero=True)(source)
encoded_source = layers.Bidirectional(
    layers.GRU(latent_dim), merge_mode="sum")(x)

**GRU-based decoder and the end-to-end model**

In [23]:
past_target = keras.Input(shape=(None,), dtype="int64", name="spanish")
x = layers.Embedding(vocab_size, embed_dim, mask_zero=True)(past_target)
decoder_gru = layers.GRU(latent_dim, return_sequences=True)
x = decoder_gru(x, initial_state=encoded_source)
x = layers.Dropout(0.5)(x)
target_next_step = layers.Dense(vocab_size, activation="softmax")(x)
seq2seq_rnn = keras.Model([source, past_target], target_next_step)

**Training our recurrent sequence-to-sequence model**

In [24]:
seq2seq_rnn.compile(
    optimizer="rmsprop",
    loss="sparse_categorical_crossentropy",
    metrics=["accuracy"])
seq2seq_rnn.fit(train_ds, epochs=15, validation_data=val_ds)

Epoch 1/15
1302/1302 [==============================] - 91s 64ms/step - loss: 1.6376 - accuracy: 0.4166 - val_loss: 1.3168 - val_accuracy: 0.5069
Epoch 2/15
1302/1302 [==============================] - 83s 64ms/step - loss: 1.3157 - accuracy: 0.5255 - val_loss: 1.1546 - val_accuracy: 0.5691
Epoch 3/15
1302/1302 [==============================] - 83s 64ms/step - loss: 1.1735 - accuracy: 0.5757 - val_loss: 1.0726 - val_accuracy: 0.6003
Epoch 4/15
1302/1302 [==============================] - 86s 66ms/step - loss: 1.0828 - accuracy: 0.6076 - val_loss: 1.0386 - val_accuracy: 0.6204
Epoch 5/15
1302/1302 [==============================] - 86s 66ms/step - loss: 1.0358 - accuracy: 0.6326 - val_loss: 1.0209 - val_accuracy: 0.6298
Epoch 6/15
1302/1302 [==============================] - 86s 66ms/step - loss: 1.0050 - accuracy: 0.6507 - val_loss: 1.0166 - val_accuracy: 0.6363
Epoch 7/15
1302/1302 [==============================] - 85s 65ms/step - loss: 0.9859 - accuracy: 0.6638 - val_loss: 1.0163 -

**Translating new sentences with our RNN encoder and decoder**

In [25]:
import numpy as np
spa_vocab = target_vectorization.get_vocabulary()
spa_index_lookup = dict(zip(range(len(spa_vocab)), spa_vocab))
max_decoded_sentence_length = 20

def decode_sequence(input_sentence):
    tokenized_input_sentence = source_vectorization([input_sentence])
    decoded_sentence = "[start]"
    for i in range(max_decoded_sentence_length):
        tokenized_target_sentence = target_vectorization([decoded_sentence])
        next_token_predictions = seq2seq_rnn.predict(
            [tokenized_input_sentence, tokenized_target_sentence])
        sampled_token_index = np.argmax(next_token_predictions[0, i, :])
        sampled_token = spa_index_lookup[sampled_token_index]
        decoded_sentence += " " + sampled_token
        if sampled_token == "[end]":
            break
    return decoded_sentence

test_eng_texts = [pair[0] for pair in test_pairs]
for _ in range(20):
    input_sentence = random.choice(test_eng_texts)
    print("-")
    print(input_sentence)
    print(decode_sequence(input_sentence))

-
I have to run.
[start] tengo que [UNK] [end]
-
There were few people in the park.
[start] había algunas personas en el parque [end]
-
I need to learn French.
[start] necesito aprender francés [end]
-
It's been a long time since we last saw each other.
[start] ha pasado mucho tiempo desde el tiempo [end]
-
You must know yourself.
[start] debes saber a ti mismo [end]
-
Tom doesn't even know how to start a lawn mower.
[start] tom ni siquiera sabe cómo [UNK] una clase de [UNK] [end]
-
Are they satisfied?
[start] están ellos [end]
-
I should've paid a little more attention.
[start] debería haber un poco más de la comida [end]
-
Tom hung out with his friends last night.
[start] tom anoche con sus amigos [end]
-
Tom wanted to buy a new pair of tennis shoes.
[start] tom quería comprar un par de zapatos por la noche de comprar [end]
-
We'll start when he gets here.
[start] cuando él te [UNK] aquí [end]
-
Tom and I have fun together.
[start] tom y yo nos juntos juntos [end]
-
Take your umbrell

### Sequence-to-sequence learning with Transformer

#### The Transformer decoder

**The `TransformerDecoder`**

In [26]:
class TransformerDecoder(layers.Layer):
    def __init__(self, embed_dim, dense_dim, num_heads, **kwargs):
        super().__init__(**kwargs)
        self.embed_dim = embed_dim
        self.dense_dim = dense_dim
        self.num_heads = num_heads
        self.attention_1 = layers.MultiHeadAttention(
            num_heads=num_heads, key_dim=embed_dim)
        self.attention_2 = layers.MultiHeadAttention(
            num_heads=num_heads, key_dim=embed_dim)
        self.dense_proj = keras.Sequential(
            [layers.Dense(dense_dim, activation="relu"),
             layers.Dense(embed_dim),]
        )
        self.layernorm_1 = layers.LayerNormalization()
        self.layernorm_2 = layers.LayerNormalization()
        self.layernorm_3 = layers.LayerNormalization()
        self.supports_masking = True

    def get_config(self):
        config = super().get_config()
        config.update({
            "embed_dim": self.embed_dim,
            "num_heads": self.num_heads,
            "dense_dim": self.dense_dim,
        })
        return config

    def get_causal_attention_mask(self, inputs):
        input_shape = tf.shape(inputs)
        batch_size, sequence_length = input_shape[0], input_shape[1]
        i = tf.range(sequence_length)[:, tf.newaxis]
        j = tf.range(sequence_length)
        mask = tf.cast(i >= j, dtype="int32")
        mask = tf.reshape(mask, (1, input_shape[1], input_shape[1]))
        mult = tf.concat(
            [tf.expand_dims(batch_size, -1),
             tf.constant([1, 1], dtype=tf.int32)], axis=0)
        return tf.tile(mask, mult)

    def call(self, inputs, encoder_outputs, mask=None):
        causal_mask = self.get_causal_attention_mask(inputs)
        if mask is not None:
            padding_mask = tf.cast(
                mask[:, tf.newaxis, :], dtype="int32")
            padding_mask = tf.minimum(padding_mask, causal_mask)
        attention_output_1 = self.attention_1(
            query=inputs,
            value=inputs,
            key=inputs,
            attention_mask=causal_mask)
        attention_output_1 = self.layernorm_1(inputs + attention_output_1)
        attention_output_2 = self.attention_2(
            query=attention_output_1,
            value=encoder_outputs,
            key=encoder_outputs,
            attention_mask=padding_mask,
        )
        attention_output_2 = self.layernorm_2(
            attention_output_1 + attention_output_2)
        proj_output = self.dense_proj(attention_output_2)
        return self.layernorm_3(attention_output_2 + proj_output)

#### Putting it all together: A Transformer for machine translation

**PositionalEmbedding layer**

In [27]:
class PositionalEmbedding(layers.Layer):
    def __init__(self, sequence_length, input_dim, output_dim, **kwargs):
        super().__init__(**kwargs)
        self.token_embeddings = layers.Embedding(
            input_dim=input_dim, output_dim=output_dim)
        self.position_embeddings = layers.Embedding(
            input_dim=sequence_length, output_dim=output_dim)
        self.sequence_length = sequence_length
        self.input_dim = input_dim
        self.output_dim = output_dim

    def call(self, inputs):
        length = tf.shape(inputs)[-1]
        positions = tf.range(start=0, limit=length, delta=1)
        embedded_tokens = self.token_embeddings(inputs)
        embedded_positions = self.position_embeddings(positions)
        return embedded_tokens + embedded_positions

    def compute_mask(self, inputs, mask=None):
        return tf.math.not_equal(inputs, 0)

    def get_config(self):
        config = super(PositionalEmbedding, self).get_config()
        config.update({
            "output_dim": self.output_dim,
            "sequence_length": self.sequence_length,
            "input_dim": self.input_dim,
        })
        return config

**End-to-end Transformer**

In [28]:
embed_dim = 256
dense_dim = 2048
num_heads = 8

encoder_inputs = keras.Input(shape=(None,), dtype="int64", name="english")
x = PositionalEmbedding(sequence_length, vocab_size, embed_dim)(encoder_inputs)
encoder_outputs = TransformerEncoder(embed_dim, dense_dim, num_heads)(x)

decoder_inputs = keras.Input(shape=(None,), dtype="int64", name="spanish")
x = PositionalEmbedding(sequence_length, vocab_size, embed_dim)(decoder_inputs)
x = TransformerDecoder(embed_dim, dense_dim, num_heads)(x, encoder_outputs)
x = layers.Dropout(0.5)(x)
decoder_outputs = layers.Dense(vocab_size, activation="softmax")(x)
transformer = keras.Model([encoder_inputs, decoder_inputs], decoder_outputs)

**Training the sequence-to-sequence Transformer**

In [29]:
transformer.compile(
    optimizer="rmsprop",
    loss="sparse_categorical_crossentropy",
    metrics=["accuracy"])
transformer.fit(train_ds, epochs=30, validation_data=val_ds)

Epoch 1/30
1302/1302 [==============================] - 76s 57ms/step - loss: 1.7063 - accuracy: 0.4056 - val_loss: 1.3620 - val_accuracy: 0.4967
Epoch 2/30
1302/1302 [==============================] - 75s 57ms/step - loss: 1.3737 - accuracy: 0.5179 - val_loss: 1.2153 - val_accuracy: 0.5483
Epoch 3/30
1302/1302 [==============================] - 77s 59ms/step - loss: 1.2161 - accuracy: 0.5707 - val_loss: 1.1035 - val_accuracy: 0.5915
Epoch 4/30
1302/1302 [==============================] - 78s 60ms/step - loss: 1.1189 - accuracy: 0.6053 - val_loss: 1.0640 - val_accuracy: 0.6135
Epoch 5/30
1302/1302 [==============================] - 78s 60ms/step - loss: 1.0636 - accuracy: 0.6299 - val_loss: 1.0240 - val_accuracy: 0.6306
Epoch 6/30
1302/1302 [==============================] - 79s 60ms/step - loss: 1.0287 - accuracy: 0.6474 - val_loss: 1.0114 - val_accuracy: 0.6380
Epoch 7/30
1302/1302 [==============================] - 79s 61ms/step - loss: 1.0034 - accuracy: 0.6618 - val_loss: 1.0000 -

**Translating new sentences with our Transformer model**

In [30]:
import numpy as np
spa_vocab = target_vectorization.get_vocabulary()
spa_index_lookup = dict(zip(range(len(spa_vocab)), spa_vocab))
max_decoded_sentence_length = 20

def decode_sequence(input_sentence):
    tokenized_input_sentence = source_vectorization([input_sentence])
    decoded_sentence = "[start]"
    for i in range(max_decoded_sentence_length):
        tokenized_target_sentence = target_vectorization(
            [decoded_sentence])[:, :-1]
        predictions = transformer(
            [tokenized_input_sentence, tokenized_target_sentence])
        sampled_token_index = np.argmax(predictions[0, i, :])
        sampled_token = spa_index_lookup[sampled_token_index]
        decoded_sentence += " " + sampled_token
        if sampled_token == "[end]":
            break
    return decoded_sentence

test_eng_texts = [pair[0] for pair in test_pairs]
for _ in range(20):
    input_sentence = random.choice(test_eng_texts)
    print("-")
    print(input_sentence)
    print(decode_sequence(input_sentence))

-
Who wants a drink?
[start] quién quiere una paz [end]
-
"What's happening in the cave? I'm curious." "I have no idea."
[start] qué ha salido de la [UNK] no tengo ni idea de salir [end]
-
He didn't attend the meeting.
[start] no [UNK] [end]
-
I believe that the boy is honest.
[start] creo que el niño es honesto [end]
-
It looks like this car is his.
[start] parece que este coche es mary [end]
-
Tom asked Mary to warm up some leftovers.
[start] tom le pidió a mary que se [UNK] algunos de los ojos [end]
-
Tom managed to tell me the truth.
[start] tom me pidió que fuera la verdad [end]
-
He's depressed.
[start] está [UNK] [end]
-
I ran home.
[start] me encontré a la casa [end]
-
The result was really satisfying.
[start] el resultado fue realmente [UNK] [end]
-
She failed to appear.
[start] ella no Él [UNK] [end]
-
You can't park here. However, there is a parking lot just around the corner.
[start] no puedes miedo por nadie [UNK] en un [UNK] por [UNK] sobre la isla [end]
-
Don't forget to

## Summary